src/NuGet.Core/NuGet.Protocol/Plugins/PluginDiscoverer.cs

```diff
@@ -140,6 +152,119 @@ private List<PluginFile> GetPluginFiles(CancellationToken cancellationToken)
+        /// The files are also expected to have a name that starts with `nuget-plugin-`
+        /// </summary>
+        /// <returns></returns>
+        private List<PluginFile> GetNetToolsPluginFiles()
+        {
+            var pluginFiles = new List<PluginFile>();
+
+            string[] paths = Array.Empty<string>();
+
+            // The path to the plugins installed using dotnet tools, should be specified in the NUGET_PLUGIN_PATHS environment variable.
+            paths = _environmentVariableReader.GetEnvironmentVariable("NUGET_PLUGIN_PATHS")?.Split(Path.PathSeparator) ?? Array.Empty<string>();
+
+            if (paths.Length == 0)
+            {
+                // If NUGET_PLUGIN_PATHS is not specified, read all the paths in PATH
+                paths = _environmentVariableReader.GetEnvironmentVariable("PATH")?.Split(Path.PathSeparator) ?? Array.Empty<string>();
+            }
+
+            foreach (var path in paths)
+            {
+                if (Directory.Exists(path))
+                {
+                    var directoryInfo = new DirectoryInfo(path);
+                    var files = directoryInfo.GetFiles("nuget-plugin-*");
+
+                    foreach (var file in files)
+                    {
+                        if (IsValidPluginFile(file))
+                        {
+                            var state = new Lazy<PluginFileState>(() => _verifier.IsValid(file.FullName) ? PluginFileState.Valid : PluginFileState.InvalidEmbeddedSignature);
+                            PluginFile pluginFile = new PluginFile(file.FullName, state, isDotnetToolsPlugin: true);
+                            pluginFiles.Add(pluginFile);
+                        }
+                    }
+                }
+            }
+
+            return pluginFiles;
+        }
+
+        /// <summary>
+        /// Checks whether a file is a valid plugin file for windows/Unix.
+        /// Windows: It should be either .bat or  .exe
+        /// Unix: It should be executable
+        /// </summary>
+        /// <param name="fileInfo"></param>
+        /// <returns></returns>
+        private static bool IsValidPluginFile(FileInfo fileInfo)
+        {
+            if (RuntimeInformation.IsOSPlatform(OSPlatform.Windows))
+            {
+                return fileInfo.Extension.Equals(".exe", StringComparison.OrdinalIgnoreCase) ||
+                       fileInfo.Extension.Equals(".bat", StringComparison.OrdinalIgnoreCase);
+            }
+            else
+            {
+#if NET8_0_OR_GREATER
+                var fileMode = File.GetUnixFileMode(fileInfo.FullName);
+
+                return fileInfo.Exists &&
+                    ((fileMode & UnixFileMode.UserExecute) != 0 ||
+                    (fileMode & UnixFileMode.GroupExecute) != 0 ||
+                    (fileMode & UnixFileMode.OtherExecute) != 0);
+#else
+                return fileInfo.Exists && (fileInfo.Attributes & FileAttributes.ReparsePoint) == 0 && IsExecutable(fileInfo);
```

Comment:
Please help me understand the reasoning behind the (fileInfo.Attributes & FileAttributes.ReparsePoint) check. My internet search suggested it is related to symbolic links, etc., but I happy to learn more.

Reply:
You're right. I wanted to ensure the file both exists and is a real file (not a symbolic link) to minimize the risk of running malicious code. For example, imagine a symbolic link located in one of the directories listed in the user's `PATH` environment variable. The link has a name like `nuget-plugin-*`, making it appear like a normal plugin, but it actually points to a malicious file in a different directory. If we run that symlink, we unintentionally execute the malicious file from the other location. We could say it is up to the user to ensure there is no malicious files in the PATH folders. However, I don't see a reason why we should leave that vulnerability open.

Reply:
It's great that you're thinking about security!  Let's talk about this offline.  I don't know how I feel about discussing security publicly.  Maybe it's fine, and a good learning resource for people, but maybe we don't want to give bad guys an insight into our security decisions.